### ISSUE TITLE
hooks for process-spawning + gcc output colorizing

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#1516]
This issue was originally created at: 2006-12-21 10:39:22.
This issue was reported by: `aviadr`.

>aviadr said at 2006-12-21 10:39:22

Hi,
 
this is concerning a patch that was already submitted to the scons-dev list on 12/14/06. 
 
this very small and concise patch (based on 0.96.93 version) that makes it easy users to create sophisticated types of `CommandActions`.  specifically, I have used it to spawn processes and retain their stdout & stderr, in order to colorize them to make the output of gcc colorful and pretty.
 
it can also be used (and indeed I will use it) for developing a class `SSHCommandAction`, which will keep an SSH session connection open and will execute commands through this connection.
 
the patch also contains example usage of this hook, that colorizes gcc output.  maybe it can also find its way into scons?  I hope someone finds these useful, and incorporate this patch into scons.
 
regards
Aviad.

>aviadr said at 2006-12-21 10:40:18

Created an attachment (id=110)
the patch

>aviadr said at 2006-12-21 10:43:08

Created an attachment (id=111)
example usage: colorizing gcc output

>stevenknight said at 2006-12-21 12:26:57

Changing this to PATCH, since FEATURE is for feature requests.
 
I'm not sure about this.  The `SPAWN` variable is already specifically intended to allow people to override how commands get spawned without having to write a subclass (we want SCons to be friendly to people who aren't savvy about OO), so this really just adds a different mechanism for doing the same thing as redefining `SPAWN`.   I think it makes more sense to have one mechanism for doing things rather than two.  If there's something you can do with the subclassing that's *functionally* different from redefining `SPAWN`, I'm not seeing it...
 
Also, I already have an `rsh.py` module on the back burner that does what you suggest with SSH (you can define ssh/rsh, etc.).  It hasn't been folded in to the code base yet because I haven't been able to figure out how to write the test scripts on Windows.  If you'd like, I'd be glad to send you a copy of what I have to see if our combined efforts could finish it off.
 
Thanks,
 
        --SK

>stevenknight said at 2007-02-09 16:40:29

Hi Aviad--
 
I've been looking at the colorization stuff and thought I'd upload what I have working (after this message since tigris.org doesn't support attachments at the same time as comments).
 
This is a prototype Tool module that would let you get GCC-colorized output by specifying it when the construction environment is initialized:
```py
    env = Environment(tools=['gcccolor'])
```
Notice that there's no need to subclass `CommandAction`, all that has to be done is to provide a function that replaces the `SPAWN` variable and captures and colorizes the output appropriately.
 
This is *not* the final version this module will take; it needs some work to make the color selection configurable, and I ultimately think it might be better as a user-settable option within the existing 'gcc' Tool module, not as a separate module.  But since I had this working I wanted to let you (and anyone else who's interested) see the in-progress work.
 
Thanks for this work; this is going to be a cool thing.
 
        --SK

>stevenknight said at 2007-02-09 16:41:38

Created an attachment (id=128)
prototype SCons Tool module for colorizing GCC output

>gregnoel said at 2008-08-05 08:01:21

Bug party triage.  SCons is moving toward using subprocess to spawn jobs, which should provide a consistent, unified, cross-platform interface.  Once that is done, it should be simpler to colorize the output.

>gregnoel said at 2008-12-26 13:20:27

Adjust triage of issues.

>stevenknight said at 2009-11-10 18:00:19

stevenknight => issues@scons

>garyo said at 2010-04-01 19:12:57

*** Issue 2584 has been marked as a duplicate of this issue. ***

>ydirson said at 2010-04-02 03:18:49

(Adding myself to CC)

>aviadr attached [spawn_hooks.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/1516/110/spawn_hooks.patch) at 2006-12-21 10:40:18.

the patch

>aviadr attached [spawn_hook.patch.example.py](https://github.com/SCons/tigris-issue-attachments/blob/master/1516/111/spawn_hook.patch.example.py) at 2006-12-21 10:43:07.

example usage: colorizing gcc output

>stevenknight attached [gcccolor.py](https://github.com/SCons/tigris-issue-attachments/blob/master/1516/128/gcccolor.py) at 2008-03-25 11:08:15.

prototype SCons Tool module for colorizing GCC output

>garyo said this issue is duplicated by #2584 at 2010-04-01 19:12:58.
